pkg/plugins/external/helpers.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package external
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"sigs.k8s.io/kubebuilder/v3/pkg/machinery"
+	"sigs.k8s.io/kubebuilder/v3/pkg/plugin/external"
+)
+
+var outputGetter ExecOutputGetter = &execOutputGetter{}
+
+// ExecOutputGetter is an interface that implements the exec output method.
+type ExecOutputGetter interface {
+	GetExecOutput(req []byte, path string) ([]byte, error)
+}
+
+type execOutputGetter struct{}
+
+func (e *execOutputGetter) GetExecOutput(request []byte, path string) ([]byte, error) {
+	cmd := exec.Command(path) //nolint:gosec
+	cmd.Stdin = bytes.NewBuffer(request)
+	cmd.Stderr = os.Stderr
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+var currentDirGetter OsWdGetter = &osWdGetter{}
+
+// OsWdGetter is an interface that implements the get current directory method.
+type OsWdGetter interface {
+	GetCurrentDir() (string, error)
+}
+
+type osWdGetter struct{}
+
+func (o *osWdGetter) GetCurrentDir() (string, error) {
+	currentDir, err := os.Getwd()
+	if err != nil {
+		return "", fmt.Errorf("error getting current directory: %v", err)
+	}
+
+	return currentDir, nil
+}
+
+func getPluginResponse(fs machinery.Filesystem, req external.PluginRequest, path string) error {
+	req.Universe = map[string]string{}
+
+	reqBytes, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	out, err := outputGetter.GetExecOutput(reqBytes, path)
+	if err != nil {
+		return err
+	}
```

Comment:
I would run all of this in the Scaffold() method of each plugin so you don't have to make a global `exec` mock.

Reply:
This is currently being run in Scaffold() method itself. [getPluginResponse](https://github.com/kubernetes-sigs/kubebuilder/pull/2338/files#diff-992e7a26b8776b12932c030601857ed56f364b1ecd7dc0fac9dcfd8c0cb903c6R39) in `Scaffold()` calls the GetExecOutput() method.

How can I avoid creating the global `exec` mock?

Reply:
```go
func (p *createAPISubcommand) Scaffold(fs machinery.Filesystem) error {
	req := external.PluginRequest{
		APIVersion: defaultAPIVersion,
		Command:    "create api",
		Args:       p.Args,
	}

	req.Universe = map[string]string{}
	reqBytes, err := json.Marshal(req)
	if err != nil {
		return err
	}

	out, err := external.GetExecOutput(reqBytes, p.Path)
	if err != nil {
		return err
	}

	wd, err := os.Getwd()
	if err != nil {
		return err
	}

	return handlePluginResponse(fs, out, wd)
```

You could also refactor the duplicated setup code into its own function.

Reply:
Thanks for the snippet @estroz. I tried this refactor and it works fine but in the unit test, I am unable to mock the `os.Getwd()` call as it's part of Scaffold now. I can mock `handlePluginResponse` by passing in test value for wd. But in the unit tests, I'm directly calling `Scaffold()` method using mock interfaces for `GetExecOutput` and `GetCurrentDir`. Any other way to have a mock value for `os.Getwd()` call?